### PR TITLE
[CA-1361] Instantiate One Tap

### PR DESCRIPTION
**Pull request by nbrr**
[Jira](https://reach5.atlassian.net/browse/CA-1361)

[PR ciam-app](https://github.com/ReachFive/ciam-app/pull/2829)
[PR reachfive-sandbox](https://github.com/ReachFive/reachfive-sandbox/pull/93)

In this PR:
- Consider google client id from the config endpoint
- New public method `instantiateOneTap` in order to load Google's scripts and initiate the One Tap widget.
- New dependency `@types/google-one-tap` for Google One Tap types.

### CODE REVIEW

File: src/main/apiClient.ts [hunk 1]
```diff
@@ -217,6 +221,52 @@ export default class ApiClient {
...
+    const script = document.createElement("script")
+    script.src = "https://accounts.google.com/gsi/client"
+    script.onload = () => this.googleOneTap(opts)
+    script.async = true
+    script.defer = true
+    document.querySelector("body")?.appendChild(script)
+  }
+
+  googleOneTap(opts: AuthOptions = {}) {
+    if (!isUndefined(this.config.googleClientId)) {
```
Comment (nbrr) [inline src/main/apiClient.ts:234]: What's the error handling in the SDK Core that could fit in the `else` of this structure?

File: src/main/oAuthClient.ts [hunk 2]
```diff
@@ -276,6 +286,52 @@ export default class OAuthClient {
...
+    return this.loginWithRedirect({
+      provider,
+      idToken,
+      nonce,
+      ...authParams
+    })
+  }
+
+  googleOneTap(opts: AuthOptions = {}) {
+    if (!isUndefined(this.config.googleClientId)) {
```
Comment (nbrr) [inline src/main/oAuthClient.ts:304]: What's the error handling in the SDK Core that could fit in the else of this structure?

File: src/main/oAuthClient.ts [hunk 3]
```diff
@@ -276,6 +286,52 @@ export default class OAuthClient {
...
+    const authParams = this.authParams({
+      ...opts,
+      useWebMessage: false
+    })
+
+    return this.loginWithRedirect({
+      provider,
+      idToken,
+      nonce,
+      ...authParams
```
Comment (StanWarz) [inline src/main/oAuthClient.ts:299]: `nonce` is already defined in `opts`, which override the ones passed in parameter.

File: src/main/oAuthClient.ts (hunk 2 above)
Comment (StanWarz) [reply src/main/oAuthClient.ts:304]: Can use `if (this.config?.googleClientId)`

Comment (StanWarz) [reply src/main/oAuthClient.ts:304]: Depends if you want the error in the console or the API response.
`logError` for console, `Promise.reject(new Error(...))` for API response.
Means you have to return a promise in the nominal case as well.

**Review by StanWarz — APPROVED** [review]

File: src/main/__tests__/startPasswordless.spec.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ beforeEach(() => {
   fetchMock.resetMocks()
 })
 
-const authParams: SingleFactorPasswordlessParams = { authType: 'magic_link', email: 'john.doe@example.com' }
+const authParams: PasswordlessParams = { authType: 'magic_link', email: 'john.doe@example.com' }
```
Comment (axuan25) [inline src/main/__tests__/startPasswordless.spec.ts:19]: Can move these changes to separate PR pls ?

File: src/main/oAuthClient.ts [hunk 5]
```diff
@@ -276,6 +286,54 @@ export default class OAuthClient {
...
+    return this.loginWithRedirect({
+      ...authParams,
+      provider,
+      idToken,
+      nonce,
+    })
+  }
+
+  private googleOneTap(opts: AuthOptions = {}): void {
+    if (this.config?.googleClientId) {
```
Comment (axuan25) [inline src/main/oAuthClient.ts:304]: Perhaps this check should be done early on in `instantiateOneTap`.

File: src/main/oAuthClient.ts [hunk 6]
```diff
@@ -276,6 +286,54 @@ export default class OAuthClient {
...
+      const nonce = randomBase64String()
+      const binaryNonce = Buffer.from(nonce, 'utf-8')
+
+      window.crypto.subtle.digest('SHA-256', binaryNonce).then(hash => {
```
Comment (axuan25) [inline src/main/oAuthClient.ts:308]: Consider returning a `Promise` here, as done for PKCE generation.

File: src/main/oAuthClient.ts [hunk 7]
```diff
@@ -276,6 +286,54 @@ export default class OAuthClient {
...
+      ...authParams,
+      provider,
+      idToken,
+      nonce,
+    })
+  }
+
+  private googleOneTap(opts: AuthOptions = {}): void {
+    if (this.config?.googleClientId) {
+      const nonce = randomBase64String()
```
Comment (axuan25) [inline src/main/oAuthClient.ts:305]: We should enable SDK users to pass their own `nonce` and only generate one if none is provided.

File: src/main/oAuthClient.ts (hunk 6 above)
Comment (axuan25) [reply src/main/oAuthClient.ts:308]: Actually, you this does return `Promise<void>`, so the error path should return a failed promise.

File: src/main/__tests__/startPasswordless.spec.ts (hunk 4 above)
Comment (nbrr) [reply src/main/__tests__/startPasswordless.spec.ts:19]: @StanWarz said he'd include this fix in a nother PR, but for now I need it on this branch so that reachfive-sandbox can run.

File: src/main/oAuthClient.ts [hunk 8]
```diff
@@ -276,6 +286,55 @@ export default class OAuthClient {
...
+      const script = document.createElement("script")
+      script.src = "https://accounts.google.com/gsi/client"
+      script.onload = () => this.googleOneTap(opts)
+      script.async = true
+      script.defer = true
+      document.querySelector("body")?.appendChild(script)
+
+      return Promise.resolve()
+    } else {
+      return Promise.reject(new Error('Google configuration missing.'))
```
Comment (axuan25) [inline src/main/oAuthClient.ts:334]: Maybe just `void` suffices here.

**Review by axuan25 — APPROVED** [review]
